File: native/android/AndroidAwnCore/src/main/kotlin/me/carda/android_awn_core/MapUtils.kt
package me.carda.android_awn_core

/**
 * Instance-based map helpers (singleton via getInstance, instance methods — no
 * static utilities): typed extraction from a `Map<String, Any?>`. Flutter
 * encodes Dart ints/bools as Int/Long/Double depending on the path, so read
 * tolerantly. Mirrors the original AndroidAwnCore type utils intent.
 */
class MapUtils private constructor() {
    companion object {
        private var instance: MapUtils? = null
        fun getInstance(): MapUtils = instance ?: MapUtils().also { instance = it }
    }

    fun getInt(value: Any?): Int? = when (value) {
        is Int -> value
        is Number -> value.toInt()
        is String -> value.toIntOrNull()
        else -> null
    }

    fun getString(value: Any?): String? = value as? String

    fun getBool(value: Any?): Boolean? = when (value) {
        is Boolean -> value
        is Number -> value.toInt() != 0
        else -> null
    }
}

File: native/apple/Sources/IosAwnCore/MapUtils.swift
import Foundation

/// Instance-based map helpers (singleton, instance methods — no static
/// utilities): typed extraction from a `[String: Any]` map. Mirrors the original
/// IosAwnCore `MapUtils` intent. Flutter encodes Dart ints/bools as Int/NSNumber
/// depending on the path, so read tolerantly.
public final class MapUtils {
    public static let shared = MapUtils()
    private init() {}

    public func getInt(_ value: Any?) -> Int? {
        if let value = value as? Int { return value }
        if let value = value as? NSNumber { return value.intValue }
        if let value = value as? String { return Int(value) }
        return nil
    }

    public func getString(_ value: Any?) -> String? {
        return value as? String
    }

    public func getBool(_ value: Any?) -> Bool? {
        if let value = value as? Bool { return value }
        if let value = value as? NSNumber { return value.boolValue }
        return nil
    }
}
